### ISSUE TITLE
any use of interpolation prohibits (??{ }) groups in the pattern

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7751]
Migrated from [rt.perl.org#33792](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=33792) (status was 'resolved')

Searchable as RT33792$

**Comment by p5pRT** [comment]
### From DanVDascalescu@yahoo.com

#### Created by DanVDascalescu@yahoo.com



This might be a documentation bug\, but several resources suggest
that it's a regex engine bug&#8203;:


perldoc perlre&#8203;:

"For reasons of security\, this construct is forbidden if the
regular expression involves run\-time interpolation of variables\,
unless the perilous use re 'eval' pragma has been used \(see the
re manpage\)\, or the variables contain results of qr// operator" \,

Mark\-Jason Dominus&#8203;:

http&#8203;://perl\.plover\.com/yak/dirty/samples/slide011\.html
http&#8203;://bumppo\.net/lists/fun\-with\-perl/2000/06/msg00061\.html



Here's the code&#8203;:

my $letter = qr/f/;
print 'foo' =~ /  \(?\{ 1 \}\)     f     /x;  \# OK
print 'foo' =~ /  \(?\{ 1 \}\)  $letter  /x;  \# ERROR\!

Although\, in accordance with perlre\, $letter \*is\* the
result of a qr// operator\, the second print generates the
following error&#8203;:

&nbsp;  Eval\-group not allowed at runtime\, use re 'eval' in regex m/ \.\.\. /


I don't think perlre means the construct is forbidden if the
variables are inside the \(?\{\.\.\.\}\) because the following is OK&#8203;:

$letter = 'x';
print 'foo' =~ /  \(?\{ $letter \}\)  /x;


I could reproduce this in 5\.6\.1\, 5\.8\.0 and 5\.8\.3\.

Regards\,
Dan Dascalescu


\-\-
Do You Yahoo?\! NO\, THANK YOU\!
Get rid of ads and banners in e\-mail with Proxmail\!
http&#8203;://google\.com/search?q=proxmail

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.3:

Configured by Debian Project at Thu Jan 29 16:05:07 EST 2004.

Summary of my perl5 (revision 5.0 version 8 subversion 3) configuration:
  Platform:
    osname=linux, osvers=2.4.22-xfs+ti1211, archname=i386-linux-thread-multi
    uname='linux kosh 2.4.22-xfs+ti1211 #1 sat oct 25 10:11:37 est 2003 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.3 -Dsitearch=/usr/local/lib/perl/5.8.3 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.3 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3.3 20040125 (prerelease) (Debian)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so.5.8.3
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.8.3:
    /etc/perl
    /usr/local/lib/perl/5.8.3
    /usr/local/share/perl/5.8.3
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.3:
    HOME=/home/knoppix
    LANG=C
    LANGUAGE=us
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/knoppix/.dist/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/X11R6/bin:/usr/local/sbin:/usr/local/bin:/usr/games:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @mauke

Hi\,

apologies in advance if you receive this mail twice\.

Anyway\, here's my test script&#8203;:

\#\!/usr/local/bin/perl
use warnings;
use strict;

my $x = qr/foo/;
my $y = 'bar';

$\_ = 'foobar';

print "case 1&#8203;: precompiled pattern\, string&#8203;: ";
if \(eval \{ m/$x$y/ \}\) \{
&nbsp;    print "ok\\n";
\} elsif \($@&#8203;\) \{
&nbsp;    print "error&#8203;: $@&#8203;";
\} else \{
&nbsp;    print "no match\\n";
\}

print "case 2&#8203;: precompiled pattern\, \(??\{ \}\) group&#8203;: ";
if \(eval \{ m/$x\(??\{$y\}\)/ \}\) \{
&nbsp;    print "ok\\n";
\} elsif \($@&#8203;\) \{
&nbsp;    print "error&#8203;: $@&#8203;";
\} else \{
&nbsp;    print "no match\\n";
\}

my $hack = qr/\(??\{$y\}\)/;
print "case 3&#8203;: precompiled pattern\, \(??\{ \}\) group inside a precompiled
pattern&#8203;: ";

if \(eval \{ m/$x$hack/ \}\) \{
&nbsp;    print "ok\\n";
\} elsif \($@&#8203;\) \{
&nbsp;    print "error&#8203;: $@&#8203;";
\} else \{
&nbsp;    print "no match\\n";
\}

\_\_END\_\_
Output&#8203;:
case 1&#8203;: precompiled pattern\, string&#8203;: ok
case 2&#8203;: precompiled pattern\, \(??\{ \}\) group&#8203;: error&#8203;: Eval\-group not allowed at
runtime\, use re 'eval' in regex m/\(?\-xism&#8203;:foo\)\(??\{$y\}\)/ at try\.pl line 20\.
case 3&#8203;: precompiled pattern\, \(??\{ \}\) group inside a precompiled pattern&#8203;: ok

As you can see\, having variables in the pattern prohibits any use of \(??\{\}\)
in the constant parts of the pattern\, even if the variables contain
precompiled
patterns\. This doesn't really make sense because you can just put the \(??\{\}\)
into another precompiled pattern\, which is accepted\.

Hmm\, on second thought I guess I understand why this happens&#8203;: the regex
string
is constructed before the regex engine gets to see it\, so any use of \(??\{\}\)
is
flagged as an error\. qr// gets compiled earlier so it can contain \(??\{\}\)
groups\. Possible fixes&#8203;: don't count precompiled patterns as interpolated \(so
/$x\(??\{"bar"\}\)/ works\)\, and/or count the number of \(??\{\}\) groups before and
after variable interpolation and croak if they differ\.

HTH\, Lukas



\-\- 
\+\+\+ Sparen Sie mit GMX DSL \+\+\+ http&#8203;://www\.gmx\.net/de/go/dsl
AKTION f???r Wechsler&#8203;: DSL\-Tarife ab 3\,99 EUR/Monat \+ Startguthaben

**Comment by p5pRT** [comment]
### From domo@ieee.org

> \[rwxr\-xr\-x@&#8203;gmx\.de \- Fri Jan 14 09&#8203;:47&#8203;:31 2005\]&#8203;:
\.\.\.
> 
> \#\!/usr/local/bin/perl
> use warnings;
> use strict;
> 
> my $x = qr/foo/;
> my $y = 'bar';
> 
> $\_ = 'foobar';
> 
> print "case 1&#8203;: precompiled pattern\, string&#8203;: ";
> if \(eval \{ m/$x$y/ \}\) \{
>     print "ok\\n";
> \} elsif \($@&#8203;\) \{
>     print "error&#8203;: $@&#8203;";
> \} else \{
>     print "no match\\n";
> \}
> 
> print "case 2&#8203;: precompiled pattern\, \(??\{ \}\) group&#8203;: ";
> if \(eval \{ m/$x\(??\{$y\}\)/ \}\) \{
>     print "ok\\n";
> \} elsif \($@&#8203;\) \{
>     print "error&#8203;: $@&#8203;";
> \} else \{
>     print "no match\\n";
> \}
> 
> my $hack = qr/\(??\{$y\}\)/;
> print "case 3&#8203;: precompiled pattern\, \(??\{ \}\) group inside a precompiled
> pattern&#8203;: ";
> 
> if \(eval \{ m/$x$hack/ \}\) \{
>     print "ok\\n";
> \} elsif \($@&#8203;\) \{
>     print "error&#8203;: $@&#8203;";
> \} else \{
>     print "no match\\n";
> \}
> 
> \_\_END\_\_
> Output&#8203;:
> case 1&#8203;: precompiled pattern\, string&#8203;: ok
> case 2&#8203;: precompiled pattern\, \(??\{ \}\) group&#8203;: error&#8203;: Eval\-group not allowed at
> runtime\, use re 'eval' in regex m/\(?\-xism&#8203;:foo\)\(??\{$y\}\)/ at try\.pl line 20\.
> case 3&#8203;: precompiled pattern\, \(??\{ \}\) group inside a precompiled pattern&#8203;: ok
> 
> As you can see\, having variables in the pattern prohibits any use of \(??\{\}\)
> in the constant parts of the pattern\, even if the variables contain
> precompiled
> patterns\. This doesn't really make sense because you can just put the \(??\{\}\)
> into another precompiled pattern\, which is accepted\.
> 
> Hmm\, on second thought I guess I understand why this happens&#8203;: the regex
> string
> is constructed before the regex engine gets to see it\, so any use of \(??\{\}\)
> is
> flagged as an error\. qr// gets compiled earlier so it can contain \(??\{\}\)
> groups\. Possible fixes&#8203;: don't count precompiled patterns as interpolated \(so
> /$x\(??\{"bar"\}\)/ works\)\, and/or count the number of \(??\{\}\) groups before and
> after variable interpolation and croak if they differ\.
> 
If the script is run with \-Mre=eval as the error message suggests\, the problem goes away\. 
Although you may be unhappy that there is a problem at all\, may I suggest that \(perhaps 
pending a real fix\) we document our way around it by incorporating your analysis into 
perlre\.pod?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
